--- crossdev
+++ crossdev
@@ -880,10 +880,17 @@
 
 	[[ ${pkg} == "[none]" ]] && return 0
 
+	case ${CTARGET} in
+		# avr requires multilib, that provides
+		# libgcc for all sub-architectures #378387
+		avr*)      set_use_force ${pkg} multilib
+		           set_use_mask ${pkg}  -multilib;;
+		*)         set_use_force ${pkg} -multilib
+		           set_use_mask ${pkg}  multilib;;
+	esac
+
 	set_keywords ${pkg} ${ver}
 	set_use ${pkg} ${use}
-	set_use_force ${pkg} -multilib
-	set_use_mask ${pkg} -multilib
 	set_links ${cat} ${pkg} "${ovl}"
 	set_env ${l} ${pkg} "${env}"
 }
