--- src/Makefile.am	2018-01-05 01:38:56.221013489 +0000
+++ src/Makefile.am	2018-01-05 01:42:12.345709205 +0000
@@ -1325,7 +1325,9 @@
 	-DCEPH_LIBDIR=\"${libdir}\" \
 	-Wnon-virtual-dtor \
 	-Wno-invalid-offsetof \
-	-Wstrict-null-sentinel
+	-Wstrict-null-sentinel \
+	-std=gnu++98
+
 # note: this is position dependant, it affects the -l options that
 # come after it on the command line. when you use ${AM_LDFLAGS} in
 # later rules, take care where you place it. for more information, see
