--- configure.in.orig	2011-02-11 20:10:23.000000000 +0100
+++ configure.in	2012-07-26 14:25:47.341404923 +0200
@@ -5,7 +5,7 @@
 AC_INIT(src/ocspd/ocspd.c)
 AC_PREREQ( 2.50)
 AC_CONFIG_AUX_DIR(build)
-AC_CONFIG_MACRO_DIR([m4])
+dnl AC_CONFIG_MACRO_DIR([m4])
 VERSION=`cat VERSION`
 
 PACKAGE_NAME="openca-ocspd"
@@ -20,7 +20,7 @@
 AC_SUBST(PKG_MIN)
 
 mybits=""
-if [[ `uname -p` = "x86_64" ]] ; then
+if [[ `uname -m` = "x86_64" ]] ; then
 	mybits="64";
 fi
 AC_SUBST(mybits)
