--- docs/Makefile.am	2012-07-28 14:15:05.724115055 +0200
+++ docs/Makefile.am	2012-07-28 14:17:59.078591399 +0200
@@ -34,27 +34,15 @@
 	$(POD2HTML) $< >$@
 
 install-man::
-	@$(mkinstalldirs) $(DESTDIR)$(exec_prefix)
-	@$(mkinstalldirs) $(DESTDIR)$(exec_prefix)/share
-	@$(mkinstalldirs) $(DESTDIR)$(exec_prefix)/share/man
-	@$(mkinstalldirs) $(DESTDIR)$(exec_prefix)/share/man/man$(MANLEV)
-	@echo "Installing Man pages is $(DESTDIR)$(exec_prefix)/share/man/man$(MANLEV) ... "
+	@$(mkinstalldirs) $(DESTDIR)$(mandir)
+	@$(mkinstalldirs) $(DESTDIR)$(mandir)/man$(MANLEV)
+	@echo "Installing Man pages is $(DESTDIR)$(mandir)/man$(MANLEV) ... "
 	@for file in $(man_MANS) ; do \
        if test -f $$file ; then \
-        $(INSTALL_DATA) $$file $(DESTDIR)$(exec_prefix)/share/man/man$(MANLEV) ; \
+        $(INSTALL_DATA) $$file $(DESTDIR)$(mandir)/man$(MANLEV) ; \
        fi ; \
     done
-	@echo "Done. Add $(DESTDIR)$(exec_prefix)/share/man to MANPATH."
-
-# install-data-am: install-man
-# 	@$(NORMAL_INSTALL) ; \
- #        $(mkinstalldirs) $(DESTDIR)$(exec_prefix)/share; \
-  #       $(mkinstalldirs) $(DESTDIR)$(exec_prefix)/share/openca-ocspd; \
-   #      for file in $(EXTRA_DIST); do \
-    #         if test -f $$file; then \
-     #          $(INSTALL_DATA) $$file $(DESTDIR)$(exec_prefix)/share/openca-ocspd; \
-      #       fi \
-       #    done
+	@echo "Done. Add $(DESTDIR)$(mandir) to MANPATH."
 
 clean:
 	@rm -rf $(man_MANS)
