--- r8168_n.c_old	2013-09-01 11:33:13.739464421 -0700
+++ r8168_n.c_new	2013-09-01 11:33:57.310919703 -0700
@@ -52,6 +52,10 @@
 #include <linux/init.h>
 #include <linux/rtnetlink.h>
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
+#include <uapi/linux/if_ether.h>
+#endif
+
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,26)
 #include <linux/pci-aspm.h>
 #endif
@@ -2232,11 +2236,13 @@
                                swab16(opts2 & 0xffff));
         ret = 0;
     }
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
+	if (opts2 & RxVlanTag)
+		__vlan_hwaccel_put_tag(skb, htons(ETH_P_8021Q), swab16(opts2 & 0xffff));
 #else
     if (opts2 & RxVlanTag)
         __vlan_hwaccel_put_tag(skb, swab16(opts2 & 0xffff));
 #endif
-
     desc->opts2 = 0;
     return ret;
 }
@@ -2294,8 +2300,11 @@
         tp->cp_cmd |= RxChkSum;
     else
         tp->cp_cmd &= ~RxChkSum;
-
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
+	if (dev->features & NETIF_F_HW_VLAN_CTAG_RX)
+#else
     if (dev->features & NETIF_F_HW_VLAN_RX)
+#endif
         tp->cp_cmd |= RxVlan;
     else
         tp->cp_cmd &= ~RxVlan;
@@ -16195,7 +16204,11 @@
 
 #ifdef CONFIG_R8168_VLAN
     if (tp->mcfg != CFG_METHOD_DEFAULT) {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
+		dev->features |= NETIF_F_HW_VLAN_CTAG_TX | NETIF_F_HW_VLAN_CTAG_RX;
+#else
         dev->features |= NETIF_F_HW_VLAN_TX | NETIF_F_HW_VLAN_RX;
+#endif
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,22)
         dev->vlan_rx_kill_vid = rtl8168_vlan_rx_kill_vid;
 #endif //LINUX_VERSION_CODE < KERNEL_VERSION(2,6,22)
@@ -16210,7 +16223,12 @@
 #else
         dev->features |= NETIF_F_RXCSUM;
         dev->hw_features = NETIF_F_SG | NETIF_F_IP_CSUM | NETIF_F_TSO |
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
+			NETIF_F_RXCSUM | NETIF_F_HW_VLAN_CTAG_TX | NETIF_F_HW_VLAN_CTAG_RX;
+#else
                            NETIF_F_RXCSUM | NETIF_F_HW_VLAN_TX | NETIF_F_HW_VLAN_RX;
+#endif
+
         dev->vlan_features = NETIF_F_SG | NETIF_F_IP_CSUM | NETIF_F_TSO |
                              NETIF_F_HIGHDMA;
 #endif
