--- r8168-8.041.01/src/r8168_n.c.orig    2015-10-06 08:33:04.000000000 +0200
+++ r8168-8.041.01/src/r8168_n.c    2016-03-20 19:38:27.755054590 +0100
@@ -4209,7 +4209,11 @@
         spin_lock_irqsave(&tp->lock, flags);
         if (dev->mtu > ETH_DATA_LEN) {
                 features &= ~NETIF_F_ALL_TSO;
-                features &= ~NETIF_F_ALL_CSUM;
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,5,0))
+features &= ~NETIF_F_CSUM_MASK;
+#else
+features &= ~NETIF_F_ALL_CSUM;
+#endif
         }
         spin_unlock_irqrestore(&tp->lock, flags);
