--- tools/libxl/Makefile.orig	2013-04-18 22:33:49.291595854 +0200
+++ tools/libxl/Makefile	2013-04-18 22:34:24.051147116 +0200
@@ -119,7 +119,7 @@
 	$(PERL) $^
 	touch $@
 
-_%.api-for-check: %.h
+_%.api-for-check: %.h $(filter-out %.api-ok, $(AUTOINCS))
 	$(CC) $(CPPFLAGS) $(CFLAGS) $(CFLAGS_$*.o) -c -E $< $(APPEND_CFLAGS) \
 		-DLIBXL_EXTERNAL_CALLERS_ONLY=LIBXL_EXTERNAL_CALLERS_ONLY \
 		>$@.new
