--- autogen.sh	2015-04-25 17:42:34.819427976 +0100
+++ autogen.sh	2015-04-25 17:41:58.459840766 +0100
@@ -146,7 +146,7 @@
 ## CHECK DEPS
 ##
 
-depends git grep
+depends grep
 
 ## BSD / GNU sed compatibility layer
 if get_path sed >/dev/null; then
@@ -182,11 +182,6 @@
     fi
 fi
 
-if ! "$git" describe --tags >/dev/null 2>&1; then
-    die "Didn't find a git repository (or no tags found). " \
-        "\`\`./autogen.sh\`\` uses git to create changelog and version information."
-fi
-
 
 ##
 ## CODE
--- autogen.sh	2015-04-25 17:49:27.651399407 +0100
+++ autogen.sh	2015-04-25 17:49:13.914889750 +0100
@@ -92,15 +92,7 @@
 
 
 get_current_version() {
-
-    version=$("$git" describe --tags)
-    if matches "$version" "$short_tag"; then
-        echo "$version"
-    else
-        version=$(echo "$version" | compat_sed "$get_short_tag")
-        echo "${version}.1dev_r$(dev_version_tag)"
-    fi
-
+    cat ../version.txt
 }
 
 prepare_files() {
--- autogen.sh	2015-04-25 17:54:30.664602733 +0100
+++ autogen.sh	2015-04-25 17:55:02.610904274 +0100
@@ -82,7 +82,7 @@
 }
 
 get_current_git_date_timestamp() {
-    "$git" show -s --pretty=format:%ct
+    cat ../datestamp.txt
 }
 
 
