--- src/openssl/pki_ocsp_resp.c	2012-07-31 21:28:23.688994060 +0200
+++ src/openssl/pki_ocsp_resp.c	2012-07-31 21:30:42.817341571 +0200
@@ -321,7 +321,6 @@
 	//	OCSP_basic_add1_cert (r->bs, issuer->cb->dup ( issuer->value ));
 
 	/* Now, if we have the otherCerts, let's add them to the response */
-	/*
 	if ( otherCerts ) {
 		int i = 0;
 		for( i = 0; i<PKI_STACK_X509_CERT_elements(otherCerts); i++ ) {
@@ -334,7 +333,6 @@
 			}
 		}
 	}
-	*/
 
 	if((PKI_X509_OCSP_RESP_DATA_sign(resp, keypair, digest)) == PKI_ERR){
 		PKI_log_err("Error while signing OCSP RESP data");
