--- lib/puppet/provider/package/portage.rb	2018-08-05 03:15:28.761080702 +0100
+++ lib/puppet/provider/package/portage.rb	2018-08-05 03:15:50.477533402 +0100
@@ -119,6 +119,8 @@
           # some fields can be empty or (null) (if we are not passed a category in the package name for instance)
           if value == '(null)'
             package_info[field] = nil
+          elsif value == '<unset>'
+            package_info[field] = nil
           elsif !value or value.empty?
             package_info[field] = nil
           else
