--- src/librc/librc.c	2012-12-07 16:53:28.000000000 +0100
+++ src/librc/librc.c	2013-04-25 20:46:25.570967776 +0200
@@ -892,12 +892,9 @@
 	list = rc_services_in_runlevel(runlevel);
 	stacks = rc_runlevel_stacks(runlevel);
 	TAILQ_FOREACH(stack, stacks, entries) {
-		sl = rc_services_in_runlevel(stack->value);
-		if (list != NULL) {
-			TAILQ_CONCAT(list, sl, entries);
-			free(sl);
-		} else
-			list = sl;
+		sl = rc_services_in_runlevel_stacked(stack->value);
+		TAILQ_CONCAT(list, sl, entries);
+		free(sl);
 	}
 	return list;
 }
