--- vgasrc/bochsvga.c   2017-05-16 17:45:13.000000000 +0200
+++ vgasrc/bochsvga.c   2017-05-16 17:47:07.000000000 +0200
@@ -56,6 +56,7 @@
     { 0x11D, { MM_DIRECT, 1600, 1200, 15, 8, 16, SEG_GRAPH } },
     { 0x11E, { MM_DIRECT, 1600, 1200, 16, 8, 16, SEG_GRAPH } },
     { 0x11F, { MM_DIRECT, 1600, 1200, 24, 8, 16, SEG_GRAPH } },
+    { 0x120, { MM_DIRECT, 480,  800,  24, 8, 16, SEG_GRAPH } },
     /* BOCHS modes */
     { 0x140, { MM_DIRECT, 320,  200,  32, 8, 16, SEG_GRAPH } },
     { 0x141, { MM_DIRECT, 640,  400,  32, 8, 16, SEG_GRAPH } },
@@ -100,6 +101,7 @@
     { 0x190, { MM_DIRECT, 1920, 1080, 16, 8, 16, SEG_GRAPH } },
     { 0x191, { MM_DIRECT, 1920, 1080, 24, 8, 16, SEG_GRAPH } },
     { 0x192, { MM_DIRECT, 1920, 1080, 32, 8, 16, SEG_GRAPH } },
+    { 0x193, { MM_DIRECT, 480,  800,  24, 8, 16, SEG_GRAPH } },
 };
 
 static int dispi_found VAR16 = 0;
